apps/ex_wire/lib/ex_wire/struct/bucket.ex

```diff
@@ -0,0 +1,273 @@
+          147, 23, 184, 239, 149, 9, 14, 119, 179, 18, 213, 204, 57, 53, 79, 134,
+          213, 214, 6>>
+      }
+  """
+  @spec tail(t()) :: Peer.t()
+  def tail(%Bucket{nodes: nodes}), do: nodes |> List.last()
+
+  @doc """
+  Inserts node to bucket.
+
+  ## Examples
+
+      iex> node = ExWire.Struct.Peer.new("13.84.180.140", 30303, "30b7ab30a01c124a6cceca36863ece12c4f5fa68e3ba9b0b51407ccc002eeed3b3102d20a88f1c1d3c3154e2449317b8ef95090e77b312d5cc39354f86d5d606", time: :test)
+      iex> ExWire.Struct.Bucket.new(1, time: :test)
+      ...> |> ExWire.Struct.Bucket.insert_node(node, time: :test)
+      %ExWire.Struct.Bucket{
+        id: 1,
+        nodes: [
+          %ExWire.Struct.Peer{
+            host: "13.84.180.140",
+            ident: "30b7ab...d5d606",
+            last_seen: 1525704921,
+            port: 30303,
+            remote_id: <<4, 48, 183, 171, 48, 160, 28, 18, 74, 108, 206, 202, 54, 134,
+              62, 206, 18, 196, 245, 250, 104, 227, 186, 155, 11, 81, 64, 124, 204, 0,
+              46, 238, 211, 179, 16, 45, 32, 168, 143, 28, 29, 60, 49, 84, 226, 68,
+              147, 23, 184, 239, 149, 9, 14, 119, 179, 18, 213, 204, 57, 53, 79, 134,
+              213, 214, 6>>
+          }
+        ],
+        updated_at: 1525704921
+      }
+  """
+  @spec insert_node(t(), Peer.t(), Keyword.t()) :: t()
+  def insert_node(bucket = %Bucket{nodes: nodes}, node, options \\ [time: :actual]) do
+    %{bucket | nodes: [node | nodes], updated_at: Timestamp.now(options[:time])}
+  end
+
+  @doc """
+  Remove node from bucket.
+
+  ## Examples
+
+      iex> node = ExWire.Struct.Peer.new("13.84.180.140", 30303, "30b7ab30a01c124a6cceca36863ece12c4f5fa68e3ba9b0b51407ccc002eeed3b3102d20a88f1c1d3c3154e2449317b8ef95090e77b312d5cc39354f86d5d606", time: :test)
+      iex> bucket = ExWire.Struct.Bucket.new(1, time: :test)
+      ...>   |> ExWire.Struct.Bucket.insert_node(node, time: :test)
+      %ExWire.Struct.Bucket{
+        id: 1,
+        nodes: [
+          %ExWire.Struct.Peer{
+            host: "13.84.180.140",
+            ident: "30b7ab...d5d606",
+            last_seen: 1525704921,
+            port: 30303,
+            remote_id: <<4, 48, 183, 171, 48, 160, 28, 18, 74, 108, 206, 202, 54, 134,
+              62, 206, 18, 196, 245, 250, 104, 227, 186, 155, 11, 81, 64, 124, 204, 0,
+              46, 238, 211, 179, 16, 45, 32, 168, 143, 28, 29, 60, 49, 84, 226, 68,
+              147, 23, 184, 239, 149, 9, 14, 119, 179, 18, 213, 204, 57, 53, 79, 134,
+              213, 214, 6>>
+          }
+        ],
+        updated_at: 1525704921
+      }
+      iex> bucket |> ExWire.Struct.Bucket.remove_node(node)
+      %ExWire.Struct.Bucket{
+         id: 1,
+         nodes: [],
+         updated_at: 1525704921
+       }
+  """
+  @spec remove_node(t(), Peer.t()) :: t()
+  def remove_node(bucket = %Bucket{nodes: nodes}, node) do
+    new_nodes =
+      nodes
+      |> Enum.drop_while(fn bucket_node ->
+        Peer.equal?(node, bucket_node)
```

Comment:
I know I've already asked before but I don't remember why we're removing all `Peer`' s here?
As I understand we have a set of unique peers. We can't have two peers with the same `remote_id` here.

Reply:
I think its ok to leave it as is for now

Reply:
I knew about this problem. But there are only 16 items in every bucket.